### PR TITLE
Document "Import From Derivation"

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
The term "Import From Derivation" is historic and does not represent the concept well.

This change does not address the adequacy of term, and only explains the concept. Reviewers: Please assess the correctness and completeness of the description and how easy it is to understand.

Renaming can be done in a follow-up PR.

This work is sponsored by [Antithesis](https://antithesis.com/) ✨

Related: https://github.com/NixOS/nix/issues/8450 (but won't fix here)

### CODE REVIEW

File: doc/manual/src/SUMMARY.md.in [hunk 1]
```diff
@@ -32,6 +32,7 @@
   - [Operators](language/operators.md)
   - [Derivations](language/derivations.md)
     - [Advanced Attributes](language/advanced-attributes.md)
+    - [Import From Derivation](language/import-from-derivation.md)
```
Comment (Ericson2314) [inline doc/manual/src/SUMMARY.md.in:35]:
```suggestion
    - [Reading build results](language/reading-build-results.md)
```

File: doc/manual/src/language/import-from-derivation.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
+# Import From Derivation
```
Comment (Ericson2314) [inline doc/manual/src/language/import-from-derivation.md:1]:
```suggestion
# Reading build results
```

**Comment by Ericson2314** [comment]
Also I still don't like "realise": it is an English word without meaning, pure nonsense jargon. I think "built" is fine; even if *we* didn't do the building *someone else* did.

Passive language can help here

- "We need to build something" OK perhaps it sounds like we need to do the building ourselves.
- "We need something built" Great, it is clear that the task can be delegated to remote builder or retrieved from a cache.
- "We need to get a build result" Great, it clear what we need, and how we get it is left open-ended as it should be.

File: doc/manual/src/language/import-from-derivation.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+It is characterised by the fact that the store object has to be [realise]d before the expression can be evaluated.
+
+[store object]: /glossary.md#gloss-store-object
+[derivation]: /glossary.md#gloss-derivation
+[realise]: /glossary.md#gloss-realise
+
+This has performance implications:
+Since evaluation is sequential, each required derivation will also be built sequentially.
+
+Passing a derivation `drv` to any built-in function that reads from the filesystem constitutes Import From Derivation:
```
Comment (Ericson2314) [inline doc/manual/src/language/import-from-derivation.md:13]: I'll use the "realize" word here as that is an orthogonal point.
```suggestion
Nix expressions can also depend on build results.

So far, we've only discussed Nix expressions that are either completely self contained, or depend on data that already exists (files on disk, something to be fetched, etc.).
But Nix expressions can also depend on build results, the [store object]s produced by [derivation]s., even if we haven't yet [realise]d that build result.
Evaluation will be paused if such a build result is needed, the build result realised, and then evaluation resumed.

[store object]: /glossary.md#gloss-store-object
[derivation]: /glossary.md#gloss-derivation
[realise]: /glossary.md#gloss-realise

This has performance implications:
Since evaluation is sequential, each required build result that the user doesn't already have in their store will also be built or downloaded sequentially.

Passing any expression that will evaluate to a [derived path] `drv^output` to any built-in function that reads from the filesystem constitutes Building-Dependent Evaluation:

[derived path]: /glossary.md#gloss-derived-path
```

File: doc/manual/src/glossary.md [hunk 4]
```diff
@@ -94,6 +186,9 @@
     derivations that are stored in the Nix store. These derivations can
     then be built.
 
+  - [IFD]{#gloss-ifd}\
+    [Import From Derivation](./language/import-from-derivation.md)
+
```
Comment (Ericson2314) [inline doc/manual/src/glossary.md:191]: Import From Derivation (IFD) can be a "historical term" that redirects to building-dependent evaluation.

File: doc/manual/src/glossary.md [hunk 5]
```diff
@@ -1,11 +1,95 @@
...
+    | Nix language       |
+    | .----------------. |
+    | | Nix expression | |
+    | '----------------' |
+    |         |          |
+    |      evaluate      |
+    |         |          |             +------------------------+
+    |         V          |             | Nix store              |
+    |   .------------.   |             |  .------------------.  |
+    |   | derivation |---|-instantiate-|->| store derivation |  |
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply doc/manual/src/glossary.md:29]:
```suggestion
    |   |  derived   |---|-points to---|->|     derivation   |  |
    |   |    path    |                 |  |  and output name |  |
```

**Comment by fricklerhandwerk** [comment]
> it is high time that we stop calling it "import from derivation"

@Ericson2314 @edolstra I agree, but renaming the concept makes this change much harder: we already have a configuration option `allow-import-from-derivation` which would need to be adapted, too. All I wanted to do is document the idea. I can do a follow-up to use a better name and a renaming strategy, which we can discuss in isolation.

>  We're not importing a derivation.

I avoided making explicit note of the historicity of the term, and focused on explaining the concept behind it instead. My request to reviewers is to check that the conceptual description is correct.

>  It is not just `import`. `builtins.readFile` and other "input ingesting"-related primops are equally involved.

[This part](https://github.com/NixOS/nix/pull/7332/files#diff-6ed75e4585860a7666b367cc189077f970429de6f551a2b6727e75b1ee153a4cR15-R22) already accounts for that fact.

> I still don't like "realise": it is an English word without meaning, pure nonsense jargon.

I closed all the comments related to https://github.com/NixOS/nix/pull/7318 and https://github.com/NixOS/nix/pull/7320. Let us discuss those in isolation, please.

File: doc/manual/src/glossary.md (hunk 4 above)
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:191]: Let us do the renaming in another pull request to focus on the essence of this one.

File: doc/manual/src/glossary.md [hunk 6]
```diff
@@ -1,11 +1,95 @@
...
+                                       |        realise         |
+                                       |           |            |
+                                       |           V            |
+                                       |    .--------------.    |
+                                       |    | store object |    |
+                                       |    '--------------'    |
+                                       +------------------------+
+    ```
+
+    [build]: #gloss-build
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:41]: Continuing that discussion [here](https://github.com/NixOS/nix/pull/7320#issuecomment-1324807476).

File: doc/manual/src/language/import-from-derivation.md (hunk 2 above)
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md:1]: Let us do the renaming in another pull request to focus on the essence of this one. See my [top-level comment](https://github.com/NixOS/nix/pull/7332#issuecomment-1324800375).

File: doc/manual/src/SUMMARY.md.in (hunk 1 above)
Comment (fricklerhandwerk) [reply doc/manual/src/SUMMARY.md.in:35]: Let us do the renaming in another pull request to focus on the essence of this one. See my [top-level comment](https://github.com/NixOS/nix/pull/7332#issuecomment-1324800375).

File: doc/manual/src/language/import-from-derivation.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    builder = /bin/sh;
+    args = [ "-c" "${/bin/sleep} 5; echo 1 > $out" ];
+    system = builtins.currentSystem;
+  };
+in import drv
+```
+
+```shellSession
+nix-instantiate IFD.nix --eval --read-write-mode
+```
```
Comment (sir4ur0n) [inline doc/manual/src/language/import-from-derivation.md:49]: This does not work for me, I think the Nix expression is wrong:
```
$ nix-instantiate ifd.nix --eval --read-write-mode
error (ignored): error: end of string reached
error: getting status of '/bin/sleep': No such file or directory
(use '--show-trace' to show detailed location information)
```

Indeed this line looks wrong:
```
args = [ "-c" "${/bin/sleep} 5; echo 1 > $out" ];
```

File: src/libexpr/eval.hh [hunk 8]
```diff
@@ -612,11 +612,9 @@ struct EvalSettings : Config
     Setting<bool> enableImportFromDerivation{
         this, true, "allow-import-from-derivation",
         R"(
-          By default, Nix allows you to `import` from a derivation, allowing
-          building at evaluation time. With this option set to false, Nix will
-          throw an error when evaluating an expression that uses this feature,
-          allowing users to ensure their evaluation will not require any
-          builds to take place.
+          By default, Nix allows [Import from Derivation](../language/import-from-derivation.md).
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply src/libexpr/eval.hh:615]: Not since https://github.com/NixOS/nix/pull/6968!

File: doc/manual/src/language/import-from-derivation.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+```nix
+# IFD.nix
+let
+  drv = derivation {
+    name = "expensive";
+    builder = /bin/sh;
+    args = [ "-c" "${/bin/sleep} 5; echo 1 > $out" ];
+    system = builtins.currentSystem;
+  };
+in import drv
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md:44]: Good idea, it indeed makes the idea a bit more obvious. I will consider this.

File: doc/manual/src/language/import-from-derivation.md (hunk 7 above)
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md:49]: Ah, that is because your system does not have `/bin/sleep`. Are you on NixOS?

I would like the manual to be self-sufficient. Importing Nixpkgs and all that stuff would make the example shorter, but also introduce lots of unrelated concepts that readers first have to understand to make sense of the example.

Unfortunately `sleep` is not a POSIX shell builtin, otherwise we could expect the default builder to just work with `sleep 5`. If you have an idea to achieve this with POSIX onboard means, that would be great.

File: src/libexpr/eval.hh (hunk 8 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply src/libexpr/eval.hh:615]: This piece will render to HTML through the build setup, which is not evident from here. But it's great that you ask, it's not evident at all from the code.

File: doc/manual/src/language/import-from-derivation.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
+# Import From Derivation
+
+Import From Derivation (IFD) is a Nix language pattern where an expression's value depends on the contents of a [store object] produced by a [derivation].
+It is characterised by the fact that the store object has to be [realise]d before the expression can be evaluated.
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md:4]: Ah, true. We have to clear that up. Because we also say "realise a store object" meaning that "a store object comes into being".

File: doc/manual/src/language/import-from-derivation.md (hunk 7 above)
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md]: Fixed by removing the sleep and instead using a string. It's not important for the principle to focus on the expensive build step.

File: doc/manual/src/glossary.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/glossary.md:29]: I agree, your proposed terminology would make things much clearer.

I think though, reworking the terminology would be quite a bit of effort where we should get together and sort this out. It would require rewriting the Derivation section and quite a bit of the glossary, too.

Therefore, even if it's not super precise, is it good enough as a first approximation?

File: doc/manual/src/language/import-from-derivation.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md]: I reworked it [here](https://github.com/NixOS/nix/pull/7320/files#diff-8327b62197c4dd0503828bcdbbea4eca2bd37528c89e52378e77536ed8140b89R15-R17) to be consistent with common usage.

**Comment by fricklerhandwerk** [comment]
I know we discussed [renaming IFD](https://github.com/NixOS/nix/issues/8450) (proposal: Read From Realisation), but let's first agree on the documentation.

File: doc/manual/src/language/import-from-derivation.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+## Example
+
+In the following Nix expression, the inner derivation `drv` produces a file containing `"hello"`.
+
+```nix
+# IFD.nix
+let
+  drv = derivation {
+    name = "hello";
+    builder = /bin/sh;
```
Comment (roberth) [inline doc/manual/src/language/import-from-derivation.md:35]:
```suggestion
    builder = "/bin/sh";
```

Don't copy `sh` from the evaluating host's FHS.
Fixes the expression for me.

File: doc/manual/src/language/import-from-derivation.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+|  '----------------'  |             |                        |
+|          |           |             |                        |
+|       evaluate       |             |                        |
+|          |           |             |                        |
+|          V           |             |                        |
+|    .------------.    |             |  .------------------.  |
+|    | derivation |----|-instantiate-|->| store derivation |  |
+|    '------------'    |             |  '------------------'  |
+|                      |             |           |            |
+|                      |             |        realise         |
```
Comment (roberth) [inline doc/manual/src/language/import-from-derivation.md:72]: If I were to read this as a UML sequence diagram, `realise` would be an internal message presumably triggered by the creation of `store derivation`, whereas in truth instantiation is a number of "add to store" operations, and "realise" is another, independent message that follows the instantiation.
(Also it would take to the store about a derivation, rather than talk to the derivation)

A UML sequence diagram would also highlight that the evaluation process exists during the realisation, whereas this diagram only shows that the Nix language exists. Maybe change the block title from "Nix language" to "Nix evaluator"?

File: src/libexpr/eval.hh [hunk 13]
```diff
@@ -835,11 +835,9 @@ struct EvalSettings : Config
...
         this, true, "allow-import-from-derivation",
         R"(
-          By default, Nix allows you to `import` from a derivation, allowing
-          building at evaluation time. With this option set to false, Nix will
-          throw an error when evaluating an expression that uses this feature,
-          allowing users to ensure their evaluation will not require any
-          builds to take place.
+          By default, Nix allows [Import from Derivation](@docroot@/language/import-from-derivation.md).
+          With this option set to `false`, Nix will throw an error when evaluating an expression that uses this feature,
+          ensuring that evaluation will not require any builds to take place.
```
Comment (roberth) [inline src/libexpr/eval.hh:840]:
```suggestion
          ensuring that the evaluated expressions will not require any builds to take place, regardless of the state of the store.
```

I think you were trying to say that it establishes a property reproducibly.

```suggestion
          even when the required store object is readily available,
          ensuring that evaluation will not require any builds to take place,
          regardless of the state of the store.
```

We could also mention pure-eval and call it reproducible, but that feels like too much info.

**Review by roberth — CHANGES_REQUESTED** [review]
- Fix the example
- A few suggestions

File: src/libexpr/eval.hh (hunk 13 above)
Comment (fricklerhandwerk) [reply src/libexpr/eval.hh:840]: Open for suggestions to mention `pure-eval`. Not sure what that would look like or how exactly it even relates.

File: doc/manual/src/language/import-from-derivation.md (hunk 12 above)
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md:73]: I'd be open to pair on fixing the diagram. Admittedly it's a bit hand-wavy but I fear it would get too noisy if done technically correct.

File: doc/manual/src/language/import-from-derivation.md [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+# IFD.nix
+let
+  drv = derivation {
+    name = "hello";
+    builder = "/bin/sh";
+    args = [ "-c" ''echo \"hello\" > $out'' ];
+    system = builtins.currentSystem;
+  };
+in "${import drv} world"
+```
```
Comment (fricklerhandwerk) [inline doc/manual/src/language/import-from-derivation.md:41]:
````suggestion
In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.

```nix
# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo hello > $out" ];
    system = builtins.currentSystem;
  };
in "${readFile drv} world"
```
````

This makes for slightly less quotation noise.

Comment (roberth) [reply doc/manual/src/language/import-from-derivation.md:41]: This adds a newline after `hello`, because of `echo`.
Also I think we should use the IFD in a subsequent derivation to prove a point later with the `nix-build` log.

How about this:

````suggestion
In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.

# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo -n hello > $out" ];
    system = builtins.currentSystem;
  };
  # import from derivation
  greeting = builtins.readFile drv;
in

derivation {
  name = builtins.replaceStrings ["h"] ["H"] greeting;
  builder = "/bin/sh";
  args = [ "-c" "echo ${greeting} > $out" ];
  system = builtins.currentSystem;
}
```
````

File: doc/manual/src/language/import-from-derivation.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+nix-instantiate IFD.nix --eval --read-write-mode
+```
+
+```
+building '/nix/store/348q1cal6sdgfxs8zqi9v8llrsn4kqkq-hello.drv'...
+"hello world"
+```
+
+Since `"hello"` is a valid Nix expression, it can be [`import`](./builtins.md#builtins-import)ed.
+That requires reading from the output [store path](@docroot@/glossary.md#gloss-store-path) of `drv`, which has to be [realised] before its contents can be read and evaluated.
```
Comment (roberth) [inline doc/manual/src/language/import-from-derivation.md:53]:
```shellSession
nix-build IFD.nix
```

```
building '/nix/store/82wwfxkqsypldrg5dgmja87n5hsgqvzz-hello.drv'...
this derivation will be built:
  /nix/store/9b89bvwrvy5lss3cqf76l69a0hywls7w-Hello.drv
building '/nix/store/9b89bvwrvy5lss3cqf76l69a0hywls7w-Hello.drv'...
/nix/store/d22adm184zl3fdliapjb44q1sdxjbhrl-Hello
```

Note in the log that `hello` had to be built before Nix could compute its final plan. This is a consequence of using IFD to construct the final derivation.

File: doc/manual/src/language/import-from-derivation.md (hunk 12 above)
Comment (roberth) [reply doc/manual/src/language/import-from-derivation.md:73]: Sgtm. Let me start with a proposal, based on my suggestion https://github.com/NixOS/nix/pull/7332#pullrequestreview-1535842400, highlighting that its derivation names aren't great yet:

```
┌───────┐     ┌─────────────┐         ┌─────────┐
│Nix CLI│     │Nix evaluator│         │Nix store│
└───┬───┘     └──────┬──────┘         └────┬────┘
    │                │                     │     
    │evaluate IFD.nix│                     │     
    │───────────────>│                     │     
    │                │                     │     
    │                │instantiate hello.drv│     
    │                │────────────────────>│     
    │                │                     │     
    │                │  realise hello.drv  │     
    │                │────────────────────>│     
    │                │                     │     
    │                │  hello.out = hello  │     
    │                │<────────────────────│     
    │                │                     │     
    │                │instantiate Hello.drv│     
    │                │────────────────────>│     
    │                │                     │     
    │ [ Hello.drv ]  │                     │     
    │<───────────────│                     │     
    │                │                     │     
    │          plan [ Hello.drv ]          │     
    │─────────────────────────────────────>│     
    │                │                     │     
    │        realise [ Hello.drv ]         │     
    │─────────────────────────────────────>│     
┌───┴───┐     ┌──────┴──────┐         ┌────┴────┐
│Nix CLI│     │Nix evaluator│         │Nix store│
└───────┘     └─────────────┘         └─────────┘

```

<details><summary>Source</summary>

https://arthursonzogni.com/Diagon/#Sequence

```
Nix CLI -> Nix evaluator: evaluate IFD.nix

Nix evaluator -> Nix store: instantiate hello.drv
Nix evaluator -> Nix store: realise hello.drv
Nix store -> Nix evaluator: hello.out = hello

Nix evaluator -> Nix store: instantiate Hello.drv
Nix evaluator -> Nix CLI: [ Hello.drv ]

Nix CLI -> Nix store: plan [ Hello.drv ]
Nix CLI -> Nix store: realise [ Hello.drv ]


```

</details>

File: doc/manual/src/language/import-from-derivation.md (hunk 15 above)
Comment (fricklerhandwerk) [reply doc/manual/src/language/import-from-derivation.md:53]: I had this originally, but it's a bit more than we really need to show, and is in general a more involved example that asks a lot of the reader to really grasp. The minimal structure is eval-build-read-eval.

Comment (roberth) [reply doc/manual/src/language/import-from-derivation.md:53]: Fair enough.

File: doc/manual/src/language/import-from-derivation.md (hunk 14 above)
Comment (roberth) [reply doc/manual/src/language/import-from-derivation.md:41]:
````suggestion
In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.

```nix
# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo -n hello > $out" ];
    system = builtins.currentSystem;
  };
in "${builtins.readFile drv} world"
```
````

Add `echo -n` to make it work as expected.

Comment (roberth) [reply doc/manual/src/language/import-from-derivation.md:41]: > Not sure if it's still as snappy as `import drv`

Not using `import` is in line with the point that it's not just about `import`, but it might also confuse, as it's not quite as clear where the IFD happens.

````suggestion
In the following Nix expression, the inner derivation `drv` produces a file containing `"hello"`.

```nix
# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo -n hello > $out" ];
    system = builtins.currentSystem;
  };
  # evaluating contents requires `drv` to be realised,
  # also known as import from derivation
  contents = builtins.readFile drv;
in "${contents} world"
```
````

**Comment by fricklerhandwerk** [comment]
This should be ready to merge.

File: doc/manual/src/language/import-from-derivation.md [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+The value of a Nix expression can depend on the contents of a [store object].
+In this case, when that store object is needed, evaluation will be paused, the store object [realised], and then evaluation resumed.
+
+[store object]: @docroot@/glossary.md#gloss-store-object
+[derivation]: @docroot@/glossary.md#gloss-derivation
+[realised]: @docroot@/glossary.md#gloss-realise
+
+This has performance implications:
+Since evaluation is sequential, each required store object that is not already in the store will also be realised sequentially.
+Usually, if store objects are not already present, realisation is orders of magnitude slower than evaluation.
```
Comment (Ericson2314) [inline doc/manual/src/language/import-from-derivation.md:12]: Not sure I quite got it but 
```suggestion
Since Nix's evaluator is sequential, Nix only discovers the store objects a Nix program requires one at a time.
If a given Nix program requires many store objects, and they are not already present realizing them one at a time is much slow than realizing them at once.  
```

1. "Orders of magnitude" means something precise, but without being concrete about what we are comparing we cannot be precise.
2. Realization itself still is parallel, just figuring out what to realize is sequential. (I still don't think my version fixes this enough.) It is an important distinction cause it means one can "batch" all the IFD at once if one is careful, which makes things a lot faster! E.g. I did that in https://github.com/obsidiansystems/obelisk/pull/1040

Obviously covering this technique is out of scope for the reference, but we should still be precise enough so that when `nix.dev` does the tutorial, the reference doesn't contradict it.

Even my "one at a time" is also wrong. Nix realizes string contexts one at a time, but a string context is a set which can refer to arbitrary many store objects. (And even if that weren't true, one could create a derivation whose only purpose is to create a store object referring to many other store objects, achieving the same outcome.)

File: doc/manual/src/language/import-from-derivation.md [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+```shellSession
+nix-instantiate IFD.nix --eval --read-write-mode
+```
+
+```
+building '/nix/store/348q1cal6sdgfxs8zqi9v8llrsn4kqkq-hello.drv'...
+"hello world"
+```
+
+The contents of the derivation's [output path](@docroot@/glossary.md#gloss-output-path) have to be [realised] before they can be read with [`readFile`](./builtins.md#builtins-readFile).
```
Comment (Ericson2314) [inline doc/manual/src/language/import-from-derivation.md:54]:
```suggestion
The contents of the derivation's [output](@docroot@/glossary.md#gloss-output) have to be [realised] before they can be read with [`readFile`](./builtins.md#builtins-readFile).
```
This is scope creep, but it would be good to change that entry. "output path" is not a great: it is just a *store path* to a *derivation output*; there is no special notion of "output path", it is just the combination of those other two concepts.

Moreover, the contents of the "output *path*" is not `hello world`, it is `asidfaosidjfoiasdfu-hello`! :grin:

**Comment by fricklerhandwerk** [comment]
Addressed all comments. Let's fix forward from here.

File: doc/manual/src/language/import-from-derivation.md [hunk 18]
```diff
@@ -0,0 +1,142 @@
+# Import From Derivation
+
+The value of a Nix expression can depend on the contents of a [store object].
+In this case, when that store object is needed, evaluation will be paused, the store object [realised], and then evaluation resumed.
```
Comment (infinisil) [inline doc/manual/src/language/import-from-derivation.md:4]: I think the first paragraph should have a short description of IFD. Currently one needs to read through 3 paragraphs to get to the first definition of IFD.

File: doc/manual/src/language/import-from-derivation.md [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+- [`builtins.readFileType`](./builtins.md#builtins-readFileType)` expr`
+- [`builtins.readDir`](./builtins.md#builtins-readDir)` expr`
+- [`builtins.pathExists`](./builtins.md#builtins-pathExists)` expr`
+- [`builtins.filterSource`](./builtins.md#builtins-filterSource)` f expr`
+- [`builtins.path`](./builtins.md#builtins-path)` { path = expr; }`
+- [`builtins.hashFile`](./builtins.md#builtins-hashFile)` t expr`
+- `builtins.scopedImport x drv`
+
+Realising store objects during evaluation can be disabled by setting [`allow-import-from-derivation`](../command-ref/conf-file.md#conf-allow-import-from-derivation) to `false`.
+Without IFD it is ensured that evaluation is complete and Nix can produce a build plan before starting any realisation.
```
Comment (infinisil) [inline doc/manual/src/language/import-from-derivation.md:30]: The IFD abbreviation isn't defined, should be in the beginning somewhere

**Review by infinisil — CHANGES_REQUESTED** [review]
Very nice! Here's some small things I noticed

**Review by infinisil — APPROVED** [review]
